include/swift/AST/TypeCheckRequests.h

```diff
@@ -2054,6 +2054,25 @@ class DifferentiableAttributeTypeCheckRequest
   void cacheResult(IndexSubset *value) const;
 };
 
+/// Checks whether a type eraser has a viable initializer.
+class TypeEraserHasViableInitRequest
+    : public SimpleRequest<TypeEraserHasViableInitRequest,
+                           bool(TypeEraserAttr *, ProtocolDecl *),
+                           CacheKind::Cached> {
+public:
+  using SimpleRequest::SimpleRequest;
+
+private:
+  friend SimpleRequest;
+
+  // Evaluation
+  llvm::Expected<bool> evaluate(Evaluator &evaluator, TypeEraserAttr *attr,
```

Comment:
Depending on what you're planning on doing with this request, it might be better if the result was the actual viable ConstructorDecl if one was found, instead of a boolean. Otherwise aren't you going to have to look up to find the candidate again?

Reply:
This is how I implemented it at first (even in the first commit of this PR), but when I extended the implementation to allow arbitrary generic requirements (as long as they're satisfied), I didn't want to re-implement overload resolution logic. I want the constraint system to handle that part

Reply:
Do you mean you're planning on building an Expr and passing it to typeCheckExpression() rather than building a fully-type checked AST for the call? I'm worried that with this approach the constraint system might pick a different overload than the one that this request picks. Also in the past we've had issues where checking this sub-constraint system could fail, which would create weird error cases that were hard to recover from. Today CSApply and ClangImporter mostly build type-checked AST for this reason.

It sounds like you already have everything you need to form the call, because you're building the SubstitutionMap for checking the generic requirements. Other than that, you would build a DeclRefExpr wrapped in a CallExpr.

But it's totally up to you, I won't press the issue further :)

Reply:
Oh, I get it, you have multiple viable candidates, and you want the constraint solver to pick one using the existing ranking rules. The attribute checker just has to guarantee that at least one viable candidate exists, not that they're unique.

Ok, that's a tricky one. Never mind.